### PR TITLE
[AIRFLOW-4509] SubDagOperator using scheduler instead of backfill

### PR DESCRIPTION
**Pull request by milton0825**
Make sure you have checked _all_ steps below.

### Jira

- [X] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-4509

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

Change `SubDagOperator` to use Airflow scheduler to schedule
tasks in subdags instead of backfill.

In the past, `SubDagOperator` relies on backfill scheduler
to schedule tasks in the subdags. Tasks in parent `DAG`
are scheduled via Airflow scheduler while tasks in
a subdag are scheduled via backfill, which complicates
the scheduling logic and adds difficulties to maintain
the two scheduling code path.

This PR simplifies how tasks in subdags are scheduled.
`SubDagOperator` is responsible for creating a DagRun for subdag
and wait until all the tasks in the subdag finish. Airflow
scheduler picks up the DagRun created by SubDagOperator,
create and schedule the tasks accordingly.

Although `SubDagOperator` can occupy a pool/concurrency slot,
user can specify the `mode=reschedule` so that the slot will be
released periodically to avoid potential deadlock.

### Tests

- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [X] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [X] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [X] Passes `flake8`

### CODE REVIEW

File: airflow/jobs/backfill_job.py
Comment (ashb) [inline airflow/jobs/backfill_job.py:269]: FYI: Docs don't matter so much on prviate (`_foo` functions) as I don't think they are rendered, but right now our docs rendering don't pick up the typing from type annotations :(

Comment (milton0825) [reply airflow/jobs/backfill_job.py:269]: Ahh that is too bad.

**Comment by milton0825** [comment]
PTAL @ashb @feng-tao @potiuk

**Comment by milton0825** [comment]
PTAL @Fokko @ashb @kaxil

**Comment by feng-tao** [comment]
@milton0825 , I don't have much experience with subdag. Hence it would be good to get someone who has that experience to review it.

cc @fenglu-g as I think composer used subdag?

**Comment by milton0825** [comment]
I can break the PR into several smaller PRs if that makes it easier to review.

**Comment by Fokko** [comment]
I'm not using subdags either. It had some performance issues in the past, and therefore I always avoided it. Most of the cases in which subdags are being used can be rewritten in such a way that it can be avoided.

**Comment by milton0825** [comment]
People were advised not to use `SubDagOperator` because of the performance issues and the bugs. I hope this PR and the following PRs can improve the user experience of `SubDagOperator` and hopefully people can start leveraging the benefits of subdags.

@Fokko do you know which committer has experience around subdags?

**Comment by Fokko** [comment]
My best guess would be @KevinYang21, otherwise @mistercrunch 👍

File: airflow/jobs/backfill_job.py
Comment (Fokko) [reply airflow/jobs/backfill_job.py:269]: Still, my IDEA loves the annotations 👍

File: airflow/jobs/scheduler_job.py
Comment (Fokko) [inline airflow/jobs/scheduler_job.py]: Now I see the annotation, this should be:
```suggestion
    def _find_dags_to_process(self, dags: List[DAG], paused_dag_ids: Set[str]):
```

Comment (Fokko) [reply airflow/jobs/scheduler_job.py]: I can also pick this up in a sepearate PR.

File: airflow/operators/subdag_operator.py
Comment (Fokko) [inline airflow/operators/subdag_operator.py]: If we explicitly pop this from the `kwargs`, then we could also add it as an argument.

File: airflow/operators/subdag_operator.py
Comment (Fokko) [inline airflow/operators/subdag_operator.py]: Instead of using the `provide_session` here, we could also use the `create_session`, the rollback logic is already there: https://github.com/apache/airflow/blob/master/airflow/utils/db.py#L32-L45

This would translate to:
```python
with create_session() as session:
    dag_run.state = State.RUNNING
    session.merge(dag_run)
    failed_task_instances = (
        session
        .query(TaskInstance)
        .filter(TaskInstance.dag_id == self.subdag.dag_id)
        .filter(TaskInstance.execution_date == execution_date)
        .filter(TaskInstance.state.in_([State.FAILED, State.UPSTREAM_FAILED]))
    )

    for task_instance in failed_task_instances:
        task_instance.state = State.NONE
        session.merge(task_instance)
```
WDYT?

File: tests/operators/test_subdag_operator.py
Comment (Fokko) [inline tests/operators/test_subdag_operator.py]: I think we should avoid using the global session, even in the tests.
```python
with create_session() as session:
        dummy_task_instance = TaskInstance(task=dummy_task, execution_date=DEFAULT_DATE, state=State.FAILED)
        session.add(dummy_task_instance)
        session.commit()
```

File: airflow/jobs/scheduler_job.py
Comment (Fokko) [inline airflow/jobs/scheduler_job.py]:
```suggestion
        paused_dag_ids = {dag.dag_id for dag in dagbag.dags.values() if dag.is_paused}
```

File: airflow/jobs/scheduler_job.py
Comment (milton0825) [reply airflow/jobs/scheduler_job.py]: Changed it to `Set`.

File: airflow/operators/subdag_operator.py
Comment (milton0825) [reply airflow/operators/subdag_operator.py]: Fixed.

File: airflow/operators/subdag_operator.py
Comment (milton0825) [reply airflow/operators/subdag_operator.py]: Yeah this is better as it wraps the `try ... finally rollback` logic in the context manager.

File: tests/operators/test_subdag_operator.py
Comment (milton0825) [reply tests/operators/test_subdag_operator.py]: Fixed.

File: airflow/jobs/scheduler_job.py
Comment (milton0825) [reply airflow/jobs/scheduler_job.py]: Fixed.

**Review by Fokko — APPROVED** [review]

File: airflow/jobs/scheduler_job.py
Comment (Fokko) [reply airflow/jobs/scheduler_job.py]: 👍

**Comment by milton0825** [comment]
It passed on my personal travis: https://travis-ci.org/milton0825/airflow/builds/562446362

**Comment by Fokko** [comment]
@milton0825 Can you do a rebase onto latest master?

**Comment by milton0825** [comment]
@Fokko just did a rebase to latest master and the build went through 🎉🎉🎉

**Comment by milton0825** [comment]
Shall we merge it?

**Comment by potiuk** [comment]
Looks like it's a nice one to merge and make it to 1.10.4 as well .. WDYT @Fokko @ashb :) ?

**Comment by Fokko** [comment]
I'm good with this one.

Since 1.10.4 is already a couple of RC underway, I would target this at 1.10.5 and 2.0.0. But this is up to @ashb :-)

**Comment by milton0825** [comment]
I am fine with releasing it in later versions. No rush.

**Comment by milton0825** [comment]
I am fine merging it but released in 1.10.5 or later 😃

**Comment by ashb** [comment]
Conflicts again, sorry @milton0825 

I don't want to pull this in to 1.10.4 on the _off_ chance that it breaks anything else (1.10.4 has already been a bit of a pain to get released) so I'll target this at 1.10.5

**Comment by milton0825** [comment]
@ashb just resolved the conflicts

**Comment by milton0825** [comment]
@ashb just resolved the conflicts

**Comment by milton0825** [comment]
Shall we merge it?

**Comment by kaxil** [comment]
Can we add `Although SubDagOperator can occupy a pool/concurrency slot,
user can specify the mode=reschedule so that the slot will be
released periodically to avoid potential deadlock.` (the line you have in PR description) as a note in the documentation of SubDagOperator?

Once that is done I'll merge it as Ash is on holiday and Fokko has already approved the PR

**Review by feng-tao — APPROVED** [review]
lgtm

**Comment by richgong** [comment]
@milton0825 @potiuk @Fokko @kaxil @ashb @feng-tao Can we get a new release with this feature included? https://github.com/apache/airflow/releases

We're using SubDags for our new experimentation pipelines at Instacart, and this would be really helpful to us. Thank you!

**Comment by mik-laj** [comment]
it will be very difficult.  Every releasr is a lot of work.it is difficult to make a separate release for one PR. Why don't you make your own private release with one change? it shouldn't be difficult.

**Comment by milton0825** [comment]
We can probably include this PR as part of our next release but I am not sure how the release timeline looks like.

**Comment by kaxil** [comment]
@richgong 

Once https://github.com/apache/airflow/pull/5743 is complete, tested, reviewed, merged and ready we plan to release 1.10.5 next month hopefully and we can include this PR then.

**Comment by Fokko** [comment]
@richgong Please not that you can install a branch or a commit using pip:
Branch: https://stackoverflow.com/questions/20101834/pip-install-from-git-repo-branch
Commit: https://stackoverflow.com/questions/13685920/install-specific-git-commit-with-pip

Instead of testing the functionality after being released, you could also test it upfront, if you hit anything then you can still create a PR to get it in 1.10.5. Hope this helps. Cheers!

**Comment by clacke** [comment]
How is this feature intended to work?

I'm finding that if I leave our sub-DAGs paused (as they were with the old style SubDagOperator), they never run, even when SubDagOperator does `run()` on them.

If I unpause our sub-DAGs, they get scheduled whenever they like, disregarding whether the main flow has reached the SubDagOperator or not. With this change it seems that sub-DAGs are just free-floating DAGs that happen to not be visible at top level.

As master and the release branches have diverged in terms of database migrations, I have cherry-picked this PR on top of 1.10.5. Are there other PRs on master I need as well in order for it to work correctly?

**Comment by milton0825** [comment]
I think it is expected that the sub-DAG does not execute when it is paused. When we pause/un-pause the parent DAG, its sub-DAGs would be paused/un-paused accordingly.

sub-DAG would only execute after its dependencies are met. It should not be free-floating like you described. Can you check if we have same issue on latest master?

**Comment by clacke** [comment]
The head of the sub-DAG itself has no explicit dependencies, the only constraint is that it should only be run when the SubDagOperator asks it to.

I will try with latest master today.

**Comment by clacke** [comment]
Tried on master today.
I trigger main DAG, all sub-DAG runs start running before their SubDagOperator is reached.

- Is it possible to make e.g. TASK3.TASK1 depend on the completion of TASK2?
 - Can I make a (subclassed) SubDagOperator xcom_push something on execute that the head of the sub-DAG can wait for?
 - Can I make SubDagOperator not create a DagRun for its sub-DAG when its DagRun is created? It seems to me that this is the core of the issue.

Other workarounds I'm considering are:
 - Stay on 1.10.5, give SubDagOperators their own big worker queue where they can sit and wait without obstructing sensors.
 - Increase number of workers without queue segregation.
 - Use disconnected DAGs and TriggerDagRunOperator instead of SubDagOperator.

**Comment by milton0825** [comment]
So a few things to clarify:
Is your parent DAG triggered by backfill or scheduler?

Via airflow scheduler, the `DagRun` of a subdag is created via a `SubDagOperator` [here](https://github.com/apache/airflow/pull/5498/files#diff-45749879e4753a355c5bdb5203584698R134) and scheduler only creates `DagRun` for non-subdags [here](https://github.com/apache/airflow/pull/5498/files#diff-c35269bcfbbe386e269ffa7487e86192R1210). So a `DagRun` of a subdag should only be created when the corresponding `SubDagOperator` runs.

Can you share your DAG and your setup so that I can take a closer look?

**Comment by clacke** [comment]
I'll construct a minimal example and try things out. The operator creating the dagrun does sound like what should sensibly happen, and matches the description of the PR, but it doesn't match my observations. I don't *think* we're doing anything weird, but a minimal example will provide better data.

**Comment by clacke** [comment]
All DAGs are Schedule: None and the parent is triggered externally with parameters using [the REST plugin](https://github.com/teamclairvoyant/airflow-rest-api-plugin), which is supposed to be equivalent to running `airflow trigger_dag` using the CLI. I don't know which version of the plugin we are using, maybe that's something to look at as well. I'm glad to hear Airflow is not *supposed* to be acting the way I'm seeing it act.

**Comment by clacke** [comment]
Correction: We have actually moved to using `POST /api/experimental/dags/:dag_id/dag_runs` for triggering the DAG, not the plugin. Would this use scheduler or backfill?

**Comment by clacke** [comment]
I can't believe I actually found it.

https://github.com/apache/airflow/blob/4b7667d2ad6e16cb3b61244c6f067ef58cfb833d/airflow/api/common/experimental/trigger_dag.py#L92

Thank you for asking good questions. I will submit a PR before long.

**Comment by milton0825** [comment]
@clacke sorry I was on vacation and was not able to reply (should have let you know).Let me know if you have a PR.

**Comment by clacke** [comment]
We have solved the problem, and uncovered another problem. Working on navigating my company's open source contribution process. Thank you for your patience. :-)

**Comment by yuqian90** [comment]
There's a lot of suggestions on the internet against using `SubDagOperator` because of deadlocking and then performance issue due to its use of `SequentialExecutor` by default.
Looks like this PR is meant to address those concerns. But from what I can see this change was not released in 1.10.6. 
Is it safe to take the `SubDagOperator` in this PR and use it as a custom operator with Airflow 1.10.6 so that the operator can be used with `CeleryExecutor` without causing deadlocks?

**Comment by ashb** [comment]
@yuqian90 Yes, it probably is safe to do that. Certainly if you try it and report back it will increas the chance that we would backport this to the next 1.10.x release.

**Pull request merged**
